highfidelity/hifi
# Optimize AvatarMixer with respect to Identity and Avatar Data

This PR achieves a couple of things:
1. It fixes any lingering parts of [FB_3234](https://highfidelity.fogbugz.com/f/cases/3234) that haven't been fixed
2. It optimizes the AvatarMixer, increasing code clarity, speed, and bandwidth usage

Known issues:
- There are certain cases where, a few seconds after opening the PAL, you MIGHT have to refresh the PAL to actually see all of the information available to you. The two known cases of this are:
1. If you open the PAL quickly after joining a domain with many avatars
2. If you open the PAL after ignoring someone, closing the PAL, and waiting 10 seconds
The reason these issues occur is because it takes a short time for the AvatarMixer to receive the information that you have opened the PAL, and know to send you data about all avatars.

**Test Plan:**
1. Navigate to dev-distributed (or another 100-avatar domain hosting a build compiled from this PR)
2. Open the PAL. You may notice that not all 100 avatars are present in the list just yet (see "Known Issues"). Still, verify that none of the avatars in the domain have frozen or disappeared.
3. Refresh the PAL. Verify that all 100 avatars are present in the PAL. Verify that none of the avatars in the domain have frozen or disappeared.
4. Ignore one of the avatars. Verify that they disappear after a moment, but their PAL sphere remains present.
5. Close the PAL. Verify that none of the avatars in the domain have frozen or disappeared.
6. Wait 20 seconds, then open the PAL.
7. You may notice that only 99 avatars are present in the PAL. Refresh the PAL.
8. Again, verify that none of the avatars in the domain have frozen or disappeared. Also verify that the PAL sphere for your ignored avatar is present.
9. Perform various other smoke tests that you can think of that involve the interaction between the PAL, the space bubble, and the "ignore" feature -- enumerating them all is impossible here.

You're done! I appreciate ya for QA testin'. :heart:
